# 18.5 - Evaluating Model Performance

### Question 1: Weather model

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import statsmodels.api as sm
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

weather = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

In [63]:
y = weather['apparenttemperature'] - weather['temperature']
X = weather[['humidity', 'windspeed']]

results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.425
Model:                            OLS   Adj. R-squared (uncentered):              0.425
Method:                 Least Squares   F-statistic:                          3.570e+04
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        14:12:35   Log-Likelihood:                     -1.7675e+05
No. Observations:               96453   AIC:                                  3.535e+05
Df Residuals:                   96451   BIC:                                  3.535e+05
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
humidity      -0.4873      0.010    -47.338      0.000      -0.507      -0.467
windspeed     -0.0772      0.001   -126.510      0.000      -0.078      -0.076
==============================================================================
Omnibus:                     9577.682   Durbin-Watson:                   0.228
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            12669.324
Skew:                          -0.867   Prob(JB):                         0.00
Kurtosis:                       3.378   Cond. No.                         27.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [64]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Wed, 20 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:12:38   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

I wasn't sure whether there was supposed to be a constant in the model. My intuition would say that it would make more sense to do so. Without the constant added in, the r-squared value is .425, and both the AIC and BIC are 3.535 * 10^5. With the constant added in, the r-squared decreases to .288. This seems surprising, for the amount of explained variance to decrease after adding in a variable, especially one as intuitive as a constant term. However, the AIC and BIC also decrease, to 3.409 * 10^5. 

Variables such as temperature and wind chill/heat index should be fairly predictable. Even an r-squared of .425 is a little disappointing.

In [65]:
weather['humidity*windspeed'] = weather['humidity']*weather['windspeed']

X = weather[['humidity', 'windspeed', 'humidity*windspeed']]

results = sm.OLS(y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.533
Model:                            OLS   Adj. R-squared (uncentered):              0.533
Method:                 Least Squares   F-statistic:                          3.677e+04
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        14:12:41   Log-Likelihood:                     -1.6670e+05
No. Observations:               96453   AIC:                                  3.334e+05
Df Residuals:                   96450   BIC:                                  3.334e+05
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
humidity               0.2820      0.011     26.590      0.000       0.261       0.303
windspeed              0.0958      0.001     74.776      0.000       0.093       0.098
humidity*windspeed    -0.3038      0.002   -149.513      0.000      -0.308      -0.300
==============================================================================
Omnibus:                     4919.327   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9471.445
Skew:                          -0.381   Prob(JB):                         0.00
Kurtosis:                       4.333   Cond. No.                         38.0
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [66]:
X = weather[['humidity', 'windspeed', 'visibility']]

results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.490
Model:                            OLS   Adj. R-squared (uncentered):              0.490
Method:                 Least Squares   F-statistic:                          3.094e+04
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        14:12:45   Log-Likelihood:                     -1.7096e+05
No. Observations:               96453   AIC:                                  3.419e+05
Df Residuals:                   96450   BIC:                                  3.419e+05
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
humidity      -1.3488      0.012   -108.590      0.000      -1.373      -1.324
windspeed     -0.1052      0.001   -167.634      0.000      -0.106      -0.104
visibility     0.0976      0.001    110.936      0.000       0.096       0.099
==============================================================================
Omnibus:                     5476.521   Durbin-Watson:                   0.283
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6619.177
Skew:                          -0.587   Prob(JB):                         0.00
Kurtosis:                       3.519   Cond. No.                         43.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Using all measures of model performance, adding in the interaction term improves performance.

Adding visibility to the model instead improves the main model, but not to the main extent that the interaction term does.

### Question 2: Housing model

In [67]:
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

house = pd.read_sql_query('select * from houseprices',con=engine)

engine.dispose()

y = house['saleprice']

X = house[['grlivarea', 'totalbsmtsf', 'fullbath', 'halfbath', 'overallqual', 'overallcond', 'yearbuilt', 'garagearea']]
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.774
Model:                            OLS   Adj. R-squared:                  0.773
Method:                 Least Squares   F-statistic:                     621.3
Date:                Wed, 20 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:13:45   Log-Likelihood:                -17458.
No. Observations:                1460   AIC:                         3.493e+04
Df Residuals:                    1451   BIC:                         3.498e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.078e+06   1.04e+05    -10.322      0.000   -1.28e+06   -8.73e+05
grlivarea      56.7268      3.607     15.727      0.000      49.651      63.802
totalbsmtsf    27.7002      3.157      8.773      0.000      21.507      33.894
fullbath    -4205.0317   2830.361     -1.486      0.138   -9757.068    1347.005
halfbath    -2034.5941   2580.532     -0.788      0.431   -7096.565    3027.377
overallqual  1.947e+04   1161.885     16.761      0.000    1.72e+04    2.18e+04
overallcond  6557.8129    986.728      6.646      0.000    4622.247    8493.379
yearbuilt     494.5793     53.807      9.192      0.000     389.031     600.128
garagearea     44.6874      6.115      7.308      0.000      32.692      56.683
==============================================================================
Omnibus:                      569.028   Durbin-Watson:                   1.979
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            83414.289
Skew:                          -0.722   Prob(JB):                         0.00
Kurtosis:                      40.002   Cond. No.                     2.93e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.93e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [68]:
X = house[['grlivarea', 'totalbsmtsf', 'fullbath', 'halfbath', 'overallqual', 'overallcond', 'yearbuilt', 'garagearea']]

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              saleprice   R-squared (uncentered):                   0.961
Model:                            OLS   Adj. R-squared (uncentered):              0.961
Method:                 Least Squares   F-statistic:                              4451.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        14:13:50   Log-Likelihood:                         -17510.
No. Observations:                1460   AIC:                                  3.504e+04
Df Residuals:                    1452   BIC:                                  3.508e+04
Df Model:                           8                                                  
Covariance Type:            nonrobust                                                  
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
grlivarea      36.8637      3.160     11.667      0.000      30.666      43.062
totalbsmtsf    35.6655      3.171     11.247      0.000      29.445      41.886
fullbath     8098.9495   2658.762      3.046      0.002    2883.525    1.33e+04
halfbath     8121.5927   2470.759      3.287      0.001    3274.954     1.3e+04
overallqual  2.385e+04   1120.544     21.281      0.000    2.16e+04     2.6e+04
overallcond  2674.5732    944.763      2.831      0.005     821.327    4527.819
yearbuilt     -59.4873      3.849    -15.456      0.000     -67.037     -51.938
garagearea     60.3864      6.134      9.844      0.000      48.353      72.420
==============================================================================
Omnibus:                      478.846   Durbin-Watson:                   1.958
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            51839.214
Skew:                          -0.469   Prob(JB):                         0.00
Kurtosis:                      32.177   Cond. No.                     7.90e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.9e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

This is another weird result with regard to the constant term. After removing the constant, the r-squared increases drastically, to .961, but the AIC and BIC increase, indicating worse performance.

In [60]:
house['paveddrive'] = house['paveddrive'].apply(lambda x: [0, 1][(x=='Y')])
house['centralair'] = house['centralair'].apply(lambda x: [0, 1][(x=='Y')])

X = house[['grlivarea', 'totalbsmtsf', 'fullbath', 'halfbath', 'overallqual', 'overallcond', 'yearbuilt', 'garagearea', 'poolarea', 'paveddrive', 'centralair']]

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              saleprice   R-squared (uncentered):                   0.961
Model:                            OLS   Adj. R-squared (uncentered):              0.961
Method:                 Least Squares   F-statistic:                              3251.
Date:                Wed, 20 Nov 2019   Prob (F-statistic):                        0.00
Time:                        14:00:05   Log-Likelihood:                         -17506.
No. Observations:                1460   AIC:                                  3.503e+04
Df Residuals:                    1449   BIC:                                  3.509e+04
Df Model:                          11                                                  
Covariance Type:            nonrobust                                                  
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
grlivarea      39.2910      3.264     12.036      0.000      32.888      45.695
totalbsmtsf    34.4983      3.212     10.741      0.000      28.198      40.798
fullbath     7377.0320   2665.975      2.767      0.006    2147.449    1.26e+04
halfbath     6804.6482   2509.013      2.712      0.007    1882.962    1.17e+04
overallqual  2.351e+04   1126.799     20.863      0.000    2.13e+04    2.57e+04
overallcond  2396.6405    961.951      2.491      0.013     509.675    4283.606
yearbuilt     -64.1610      4.189    -15.315      0.000     -72.379     -55.943
garagearea     57.2037      6.231      9.181      0.000      44.982      69.426
poolarea      -32.1155     26.070     -1.232      0.218     -83.254      19.023
paveddrive   7588.7371   4127.398      1.839      0.066    -507.577    1.57e+04
centralair   7095.9417   4664.184      1.521      0.128   -2053.333    1.62e+04
==============================================================================
Omnibus:                      440.318   Durbin-Watson:                   1.962
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            46887.509
Skew:                          -0.276   Prob(JB):                         0.00
Kurtosis:                      30.757   Cond. No.                     1.34e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.34e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [61]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.775
Model:                            OLS   Adj. R-squared:                  0.773
Method:                 Least Squares   F-statistic:                     452.8
Date:                Wed, 20 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:00:31   Log-Likelihood:                -17456.
No. Observations:                1460   AIC:                         3.494e+04
Df Residuals:                    1448   BIC:                         3.500e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.147e+06   1.14e+05    -10.086      0.000   -1.37e+06   -9.24e+05
grlivarea      57.7585      3.649     15.828      0.000      50.600      64.917
totalbsmtsf    28.2431      3.167      8.918      0.000      22.031      34.455
fullbath    -4981.8011   2854.290     -1.745      0.081   -1.06e+04     617.184
halfbath    -2162.1771   2583.871     -0.837      0.403   -7230.708    2906.354
overallqual  1.932e+04   1165.908     16.573      0.000     1.7e+04    2.16e+04
overallcond  7008.6778   1036.488      6.762      0.000    4975.499    9041.856
yearbuilt     531.7788     59.225      8.979      0.000     415.603     647.955
garagearea     45.0729      6.144      7.337      0.000      33.022      57.124
poolarea      -39.3449     25.218     -1.560      0.119     -88.814      10.124
paveddrive  -2912.5434   4124.591     -0.706      0.480    -1.1e+04    5178.269
centralair  -5154.5991   4670.755     -1.104      0.270   -1.43e+04    4007.572
==============================================================================
Omnibus:                      533.870   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            76879.045
Skew:                          -0.595   Prob(JB):                         0.00
Kurtosis:                      38.530   Cond. No.                     3.19e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.19e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Adding in pool square footage, whether a house has a paved driveway, and whether a house has central air conditioning does not improve our model. The adjusted r-squared values remain the same, and 2 of the 4 AIC/BIC values get worse, with one improving and one staying the same.